### PR TITLE
Allow picking of a random subfaction

### PR DESCRIPTION
**Pull request by abcdefg30**
The "https://github.com/OpenRA/OpenRA/pull/7399#issuecomment-74219328 version" of #7399.

### CODE REVIEW

**Comment by abcdefg30** [comment]
Added two flags for RandomAllies and RandomSoviet.

File: OpenRA.Game/Player.cs [hunk 1]
```diff
@@ -52,8 +52,19 @@ static CountryInfo ChooseCountry(World world, string name)
 				.WithInterface<CountryInfo>().Where(c => c.Selectable)
 				.ToList();
 
-			return selectableCountries.FirstOrDefault(c => c.Race == name)
+			var selected = selectableCountries.FirstOrDefault(c => c.Race == name)
```
Comment (Phrohdoh) [inline OpenRA.Game/Player.cs]: What if:
```
Country@rands:
    Race: Randoms
    RandomRace: Rand1, Rand2

Country@r1
    Race: Rand1
    RandomRace: <other randoms>
```

etc.

File: mods/d2k/rules/world.yaml [hunk 2]
```diff
@@ -94,6 +94,10 @@ World:
 		OverrideFullShroud: full
 		OverrideFullFog: full
 		ShroudBlend: Multiply
+	Country@Random:
+		Name: Random
```
Comment (Phrohdoh) [inline mods/d2k/rules/world.yaml]: I think we should use either `Random` or `Any` in all mods.

File: OpenRA.Game/Player.cs (hunk 1 above)
Comment (abcdefg30) [inline OpenRA.Game/Player.cs]: > 14:17 | (+pchote) Is this the comment about modders intentionally specifying bogus races?
14:17 | (+Phrohdoh) pchote: it’s not bogus, its multiple levels of random
14:17 | (+pchote) abcdefg30: this isnt a problem that needs solving

File: mods/d2k/rules/world.yaml (hunk 2 above)
Comment (abcdefg30) [inline mods/d2k/rules/world.yaml]: Using ```Any``` now.

**Comment by abcdefg30** [comment]
Updated.

**Comment by pchote** [comment]
The code changes look fine, and it works ingame.

The one thing I don't like is that the "Any Soviet" label is too long, and runs off the side of the list.  The simple option would be to drop the "Any", and rely on the Random grouping and the logo to differentiate between Soviet (country) and Soviet (random).

**Comment by Phrohdoh** [comment]
Not a fan of these flags tbh, they look messy.

File: OpenRA.Game/Player.cs [hunk 3]
```diff
@@ -52,8 +52,19 @@ static CountryInfo ChooseCountry(World world, string name)
...
 				?? selectableCountries.Random(world.SharedRandom);
+
+			if (selected.RandomRace.Any())
```
Comment (obrakmann) [inline OpenRA.Game/Player.cs]: You could solve (most of) the problem Phrohdoh mentioned immediately above by changing this `if` to `while`. However, if the modder screws it up and creates a loop, you'd get an infinite loop here. On the plus side, you'd get support for multiple levels of random races.

**Comment by obrakmann** [comment]
Code looks good and works as expected. Up to you if you want to adopt my suggestion.

I agree with Phrohdoh that the flags don't work so well. The question mark is hardly discernible. Either make the question mark broader, or maybe use the flags from Phrohdoh's original PR.

**Comment by abcdefg30** [comment]
Updated.

**Comment by obrakmann** [comment]
Code looks good, and the random selection works as expected both in ra and d2k. :+1:

File: OpenRA.Game/Traits/World/Country.cs [hunk 4]
```diff
@@ -18,6 +18,9 @@ public class CountryInfo : TraitInfo<Country>
 		[Desc("This is the internal name for owner checks.")]
 		public readonly string Race = null;
 
+		[Desc("Pick a random race as the player's race out of this list.")]
+		public readonly string[] RandomRace = { };
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/World/Country.cs]: I'd be much happier if this were `RandomRaceMembers` or `RandomMembers`, atm this sounds like a bool.

Comment (abcdefg30) [inline OpenRA.Game/Traits/World/Country.cs]: I went for `RandomRaceMembers`.

**Comment by abcdefg30** [comment]
Updated.

**Comment by Phrohdoh** [comment]
@Mailaender can you restart AppVeyor please?

**Comment by Mailaender** [comment]
Okay. You should also be able to.

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 5]
```diff
@@ -270,7 +270,7 @@ public bool InterpretCommand(S server, Connection conn, Session.Client client, s
 								Name = botType,
 								Bot = botType,
 								Slot = parts[0],
-								Country = "random",
+								Race = "Random",
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: I'm going to nitpick once more, could we keep this as `random` please? For consistency.

**Comment by Phrohdoh** [comment]
:+1: works for me in our mods and RA2. Merging, thanks @abcdefg30 for picking this up!

**Comment by Phrohdoh** [comment]
[Changelog entry](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/fb7c34a7aa4647b6280a9d5622b29ec6dbc57008...cc874a18c4087630592e11d8dff96d2a4445d6ba).

**Pull request merged**